File: fixtures/vulnerable-app/src/agent.ts
// FIXTURE — planted unsafe agent/tool execution. A model-produced function argument
// reaches a Node shell string through one local wrapper with no approval or allowlist.
import { exec } from "node:child_process";
import OpenAI from "openai";

const openai = new OpenAI();

function runToolCommand(command: string) {
  return exec(command);
}

export async function runAgent() {
  const completion = await openai.chat.completions.create({
    model: "gpt-4o",
    messages: [{ role: "user", content: "Inspect this repository" }],
    tools: [{ type: "function", function: { name: "run_command", parameters: {} } }],
  });
  const toolCall = completion.choices[0].message.tool_calls[0];
  const args = JSON.parse(toolCall.function.arguments);
  return runToolCommand(args.command);
}

File: fixtures/ruby-ai-corpus/fixed/agent.rb
require "json"

TOOLS = {
  "read_file" => ->(arguments) { File.read(arguments.fetch("path")) },
  "list_files" => ->(arguments) { Dir.children(arguments.fetch("path", ".")) },
}.freeze

def execute_tool(tool_call)
  arguments = JSON.parse(tool_call.function.arguments)
  tool = TOOLS.fetch(tool_call.function.name)
  tool.call(arguments)
end
